File: extensions/dashboard/contract/shell/src/lib/tokens.ts
/**
 * Token → Tailwind translation layer.
 *
 * The Go contract emits string tokens (e.g., "4" for spacing, "center" for
 * alignment) that map to Tailwind utility classes. Keeping this mapping in
 * one place enforces the rule that only canonical shadcn semantic tokens
 * leak into rendered class names — no success / warning / info, no custom
 * brand hex codes.
 */

export type Spacing =
  | "0"
  | "0.5"
  | "1"
  | "1.5"
  | "2"
  | "2.5"
  | "3"
  | "4"
  | "5"
  | "6"
  | "8"
  | "10"
  | "12"
  | "16"
  | "20"
  | "24";

export type Align = "start" | "center" | "end" | "stretch" | "baseline";
export type Justify =
  | "start"
  | "center"
  | "end"
  | "between"
  | "around"
  | "evenly";

export type Direction = "row" | "column" | "row-reverse" | "column-reverse";
export type Orientation = "horizontal" | "vertical";
export type ContainerSize = "sm" | "md" | "lg" | "xl" | "2xl" | "full";
export type Radius = "none" | "sm" | "md" | "lg" | "xl" | "full";
export type ColorVariant =
  | "default"
  | "secondary"
  | "destructive"
  | "outline"
  | "ghost"
  | "link"
  | "muted"
  | "accent";

export type TextSize = "xs" | "sm" | "base" | "lg" | "xl";
export type TextVariant =
  | "default"
  | "muted"
  | "destructive"
  | "primary"
  | "secondary";
export type Size = "xs" | "sm" | "default" | "lg" | "xl" | "icon";

export interface ResponsiveCols {
  base?: number;
  sm?: number;
  md?: number;
  lg?: number;
  xl?: number;
  "2xl"?: number;
}

export function gapClass(s: Spacing | undefined): string {
  if (!s) return "";
  return `gap-${s}`;
}

export function rowGapClass(s: Spacing | undefined): string {
  return s ? `gap-y-${s}` : "";
}

export function colGapClass(s: Spacing | undefined): string {
  return s ? `gap-x-${s}` : "";
}

export function paddingClass(s: Spacing | undefined): string {
  return s ? `p-${s}` : "";
}

export function alignClass(a: Align | undefined): string {
  if (!a) return "";
  return {
    start: "items-start",
    center: "items-center",
    end: "items-end",
    stretch: "items-stretch",
    baseline: "items-baseline",
  }[a];
}

export function justifyClass(j: Justify | undefined): string {
  if (!j) return "";
  return {
    start: "justify-start",
    center: "justify-center",
    end: "justify-end",
    between: "justify-between",
    around: "justify-around",
    evenly: "justify-evenly",
  }[j];
}

export function directionClass(d: Direction | undefined): string {
  if (!d) return "";
  return {
    row: "flex-row",
    column: "flex-col",
    "row-reverse": "flex-row-reverse",
    "column-reverse": "flex-col-reverse",
  }[d];
}

export function containerClass(s: ContainerSize | undefined): string {
  if (!s) return "max-w-7xl";
  return {
    sm: "max-w-screen-sm",
    md: "max-w-screen-md",
    lg: "max-w-screen-lg",
    xl: "max-w-screen-xl",
    "2xl": "max-w-screen-2xl",
    full: "max-w-full",
  }[s];
}

export function radiusClass(r: Radius | undefined): string {
  if (!r) return "";
  return {
    none: "rounded-none",
    sm: "rounded-sm",
    md: "rounded-md",
    lg: "rounded-lg",
    xl: "rounded-xl",
    full: "rounded-full",
  }[r];
}

export function textSizeClass(s: TextSize | undefined): string {
  if (!s) return "";
  return { xs: "text-xs", sm: "text-sm", base: "text-base", lg: "text-lg", xl: "text-xl" }[s];
}

export function textVariantClass(v: TextVariant | undefined): string {
  if (!v) return "";
  return {
    default: "text-foreground",
    muted: "text-muted-foreground",
    destructive: "text-destructive",
    primary: "text-primary",
    secondary: "text-secondary-foreground",
  }[v];
}

export function gridColsClass(cols: ResponsiveCols | number | undefined): string {
  if (cols === undefined || cols === null) return "";
  if (typeof cols === "number") return colsForN(cols, "");
  const out: string[] = [];
  if (cols.base !== undefined) out.push(colsForN(cols.base, ""));
  if (cols.sm !== undefined) out.push(colsForN(cols.sm, "sm:"));
  if (cols.md !== undefined) out.push(colsForN(cols.md, "md:"));
  if (cols.lg !== undefined) out.push(colsForN(cols.lg, "lg:"));
  if (cols.xl !== undefined) out.push(colsForN(cols.xl, "xl:"));
  if (cols["2xl"] !== undefined) out.push(colsForN(cols["2xl"], "2xl:"));
  return out.filter(Boolean).join(" ");
}

function colsForN(n: number, prefix: string): string {
  if (n <= 0) return "";
  if (n > 12) return `${prefix}grid-cols-12`;
  return `${prefix}grid-cols-${n}`;
}

File: extensions/dashboard/contract/components/tokens.go
package components

// ColorVariant names the shadcn semantic palette. Only canonical tokens
// are exposed — there is no Success, Warning, or Info to keep the contract
// strictly shadcn-aligned. Map non-shadcn semantics to the closest token:
// errors → Destructive, warnings → Muted, info/notices → Accent.
type ColorVariant string

const (
	VariantDefault     ColorVariant = "default"
	VariantSecondary   ColorVariant = "secondary"
	VariantDestructive ColorVariant = "destructive"
	VariantOutline     ColorVariant = "outline"
	VariantGhost       ColorVariant = "ghost"
	VariantLink        ColorVariant = "link"
	VariantMuted       ColorVariant = "muted"
	VariantAccent      ColorVariant = "accent"
)

// Size is the shared sizing scale for buttons, badges, inputs, avatars.
type Size string

const (
	SizeSm      Size = "sm"
	SizeDefault Size = "default"
	SizeLg      Size = "lg"
	SizeIcon    Size = "icon"
	SizeXs      Size = "xs"
	SizeXl      Size = "xl"
)

// Spacing maps to Tailwind's spacing scale (0, 0.5, 1, 1.5, ..., 24).
// Use the string form to avoid coupling Go callers to Tailwind's number
// alphabet — the React renderer interprets it as a Tailwind gap utility.
type Spacing string

const (
	SpacingNone Spacing = "0"
	SpacingXs   Spacing = "1"
	SpacingSm   Spacing = "2"
	SpacingMd   Spacing = "4"
	SpacingLg   Spacing = "6"
	SpacingXl   Spacing = "8"
	Spacing2xl  Spacing = "10"
)

// Align maps to items-* (flex cross-axis alignment).
type Align string

const (
	AlignStart    Align = "start"
	AlignCenter   Align = "center"
	AlignEnd      Align = "end"
	AlignStretch  Align = "stretch"
	AlignBaseline Align = "baseline"
)

// Justify maps to justify-* (flex main-axis alignment).
type Justify string

const (
	JustifyStart   Justify = "start"
	JustifyCenter  Justify = "center"
	JustifyEnd     Justify = "end"
	JustifyBetween Justify = "between"
	JustifyAround  Justify = "around"
	JustifyEvenly  Justify = "evenly"
)

// Direction is the flex axis.
type Direction string

const (
	DirectionRow      Direction = "row"
	DirectionColumn   Direction = "column"
	DirectionRowRev   Direction = "row-reverse"
	DirectionColumnRv Direction = "column-reverse"
)

// Radius maps to Tailwind's rounded-* utilities.
type Radius string

const (
	RadiusNone Radius = "none"
	RadiusSm   Radius = "sm"
	RadiusMd   Radius = "md"
	RadiusLg   Radius = "lg"
	RadiusXl   Radius = "xl"
	RadiusFull Radius = "full"
)

// Shadow maps to Tailwind's shadow-* utilities.
type Shadow string

const (
	ShadowNone Shadow = "none"
	ShadowSm   Shadow = "sm"
	ShadowMd   Shadow = "md"
	ShadowLg   Shadow = "lg"
	ShadowXl   Shadow = "xl"
)

// Orientation distinguishes horizontal/vertical for separators, splits,
// radio groups, accordions.
type Orientation string

const (
	OrientationHorizontal Orientation = "horizontal"
	OrientationVertical   Orientation = "vertical"
)

// ContainerSize is the max-width preset for layout.container.
type ContainerSize string

const (
	ContainerSm   ContainerSize = "sm"
	ContainerMd   ContainerSize = "md"
	ContainerLg   ContainerSize = "lg"
	ContainerXl   ContainerSize = "xl"
	Container2xl  ContainerSize = "2xl"
	ContainerFull ContainerSize = "full"
)

// ResponsiveCols is the grid column count. Allows either a fixed integer
// or a per-breakpoint object.
type ResponsiveCols struct {
	Base int `json:"base,omitempty"`
	Sm   int `json:"sm,omitempty"`
	Md   int `json:"md,omitempty"`
	Lg   int `json:"lg,omitempty"`
	Xl   int `json:"xl,omitempty"`
	X2xl int `json:"2xl,omitempty"`
}

// Cols returns a ResponsiveCols with the same column count across all
// breakpoints. Use the struct literal directly when you need responsive
// column counts.
func Cols(n int) ResponsiveCols {
	return ResponsiveCols{Base: n}
}

// HeadingLevel is the semantic level of a heading element (1–6).
type HeadingLevel int

// AspectRatio maps to Tailwind's aspect-* utilities.
type AspectRatio string

const (
	AspectSquare AspectRatio = "square"
	AspectVideo  AspectRatio = "video"
	AspectAuto   AspectRatio = "auto"
)

// ObjectFit maps to Tailwind's object-* utilities.
type ObjectFit string

const (
	ObjectCover   ObjectFit = "cover"
	ObjectContain ObjectFit = "contain"
	ObjectFill    ObjectFit = "fill"
	ObjectNone    ObjectFit = "none"
)

// TextSize is the typographic size for atom.text.
type TextSize string

const (
	TextXs   TextSize = "xs"
	TextSm   TextSize = "sm"
	TextBase TextSize = "base"
	TextLg   TextSize = "lg"
	TextXl   TextSize = "xl"
)

// TextVariant is the semantic color tone for atom.text. Maps to shadcn
// foreground tokens.
type TextVariant string

const (
	TextVariantDefault     TextVariant = "default"
	TextVariantMuted       TextVariant = "muted"
	TextVariantDestructive TextVariant = "destructive"
	TextVariantPrimary     TextVariant = "primary"
	TextVariantSecondary   TextVariant = "secondary"
)
